File: android/src/main/kotlin/com/gvuitech/mediax/MediaxPlugin.kt
package com.gvuitech.mediax

import android.app.Activity
import android.app.Application.ActivityLifecycleCallbacks
import android.app.PictureInPictureParams
import android.content.Context
import android.content.res.Configuration
import android.os.Build
import android.os.Bundle
import android.util.Log
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.platform.PlatformViewFactory
import io.flutter.plugin.platform.PlatformView
import io.flutter.embedding.engine.dart.DartExecutor
import io.flutter.embedding.engine.loader.FlutterLoader

/** MediaxPlugin */
class MediaxPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

  private lateinit var context: Context
  private lateinit var channel: MethodChannel
  private lateinit var flutterPluginBinding: FlutterPlugin.FlutterPluginBinding
  private var activity: Activity? = null

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    this.context = flutterPluginBinding.applicationContext

    this.flutterPluginBinding = flutterPluginBinding
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, Constants.MEDIAX)
    channel.setMethodCallHandler(this)

    val platformViewRegistry = flutterPluginBinding.platformViewRegistry

    // Register the platform view factory
    platformViewRegistry.registerViewFactory(
      Constants.VIDEO_VIEW,
      VideoPlayerViewFactory(flutterPluginBinding.flutterEngine)
    )
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      Constants.INIT_PLAYER -> {
        val controllerId = (call.arguments as? Map<*, *>)?.get(Constants.CONTROLLER_ID) as? String
        controllerId?.let {
          val controllerChannel = MethodChannel(
            flutterPluginBinding.binaryMessenger,
            "mediax_$it"
          )
          if (ControllerManager.getController(it) == null) {
            ControllerManager.createController(
              it,
              flutterPluginBinding.flutterEngine,
              context,
              activity,
              call.arguments as Map<String?, Any?>,
              controllerChannel
            )
          }
        }
        result.success(null)
      }
      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    if (::channel.isInitialized) {
      channel.setMethodCallHandler(null)
    }

    ControllerManager.releaseAllControllers()

    if (::flutterPluginBinding.isInitialized) {
      flutterPluginBinding = binding
    }
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
      activity = binding.activity
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivity() {
    activity = null
  }
}

File: macos/Classes/MediaxPlugin.swift
import Cocoa
import FlutterMacOS
import AVFoundation

public class MediaxPlugin: NSObject, FlutterPlugin {
    private static var pluginRegistrar: FlutterPluginRegistrar?
    
  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = MediaxPlugin()
        pluginRegistrar = registrar
        
        // Handle controller creation
        let channel = FlutterMethodChannel(name: Constants.MEDIAX, binaryMessenger: registrar.messenger)
    registrar.addMethodCallDelegate(instance, channel: channel)
        
        // Register the view factory
        let factory = PlayerViewFactory(registrar: registrar)
        registrar.register(factory, withId: Constants.VIDEO_VIEW)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
        case Constants.INIT_PLAYER:
            // Retrieve controllerId from arguments
            if let arguments = call.arguments as? [String: Any],
               let controllerId = arguments[Constants.CONTROLLER_ID] as? String {
                
                // Create a method channel for this specific controller
                let controllerChannel = FlutterMethodChannel(
                    name: "mediax_\(controllerId)",
                    binaryMessenger: MediaxPlugin.pluginRegistrar!.messenger
                )
                
                // Check if the controller exists, if not create it
                if ControllerManager.getController(controllerId: controllerId) == nil {
                    do {
                        try ControllerManager.createController(
                            controllerId: controllerId,
                            params: arguments,
                            methodChannel: controllerChannel,
                            registrar: MediaxPlugin.pluginRegistrar!
                        )
                    } catch {
                        print("Error creating controller: \(error)")
                    }
                }
            }
            result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
    
    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        ControllerManager.releaseAllControllers()
    }
}
